sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ReduceNumShufflePartitions.scala

```diff
@@ -53,17 +50,40 @@ import org.apache.spark.internal.Logging
  *  - post-shuffle partition 2: pre-shuffle partition 2 (size 170 MiB)
  *  - post-shuffle partition 3: pre-shuffle partition 3 and 4 (size 50 MiB)
  */
-class ExchangeCoordinator(
-    advisoryTargetPostShuffleInputSize: Long,
-    minNumPostShufflePartitions: Int = 1)
-  extends Logging {
+case class ReduceNumShufflePartitions(conf: SQLConf) extends Rule[SparkPlan] {
+
+  override def apply(plan: SparkPlan): SparkPlan = {
+    val shuffleMetrics: Seq[MapOutputStatistics] = plan.collect {
+      case stage: ShuffleQueryStage =>
+        val metricsFuture = stage.mapOutputStatisticsFuture
+        assert(metricsFuture.isCompleted, "ShuffleQueryStage should already be ready")
+        ThreadUtils.awaitResult(metricsFuture, Duration.Zero)
+    }.filter(_ != null) // ShuffleQueryStage may give null mapOutputStatistics, skip it.
+
+    if (shuffleMetrics.nonEmpty) {
```

Comment:
When bucket table joins non-bucket table, it could fail.

For example:
```
sql("drop table bucketed_table1").collect
sql("drop table table2").collect

val df1 = (0 until 50).map(i => (i % 5, i % 13, i.toString)).toDF("i", "j", "k").as("df1")
val df2 = (0 until 50).map(i => (i % 7, i % 11, i.toString)).toDF("i", "j", "k").as("df2")
df1.write.format("parquet").bucketBy(20, "i").saveAsTable("bucketed_table1")
df2.write.format("parquet").saveAsTable("table2")

spark.conf.set("spark.sql.autoBroadcastJoinThreshold", "0")
spark.conf.set("spark.sql.adaptive.maxNumPostShufflePartitions", 10)
sql("select * from bucketed_table1 t1 join table2 t2 on t1.i = t2.i").collect()
```

```
java.lang.IllegalArgumentException: Can't zip RDDs with unequal numbers of partitions: List(20, 1)
  at org.apache.spark.rdd.ZippedPartitionsBaseRDD.getPartitions(ZippedPartitionsRDD.scala:58)
  at org.apache.spark.rdd.RDD.$anonfun$partitions$2(RDD.scala:254)
  at scala.Option.getOrElse(Option.scala:138)
  at org.apache.spark.rdd.RDD.partitions(RDD.scala:252)
  at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:49)
  at org.apache.spark.rdd.RDD.$anonfun$partitions$2(RDD.scala:254)
  at scala.Option.getOrElse(Option.scala:138)
  at org.apache.spark.rdd.RDD.partitions(RDD.scala:252)
  at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:49)
  at org.apache.spark.rdd.RDD.$anonfun$partitions$2(RDD.scala:254)
```